vyper/semantics/analysis/import_graph.py

```diff
@@ -0,0 +1,35 @@
+import contextlib
+from dataclasses import dataclass, field
+from typing import Iterator
+
+from vyper import ast as vy_ast
+from vyper.exceptions import ImportCycle
+
+"""
+data structure for collecting import statements and validating the
+import graph
+"""
+
+
+@dataclass
+class ImportGraph:
+    # the current path in the import graph traversal
+    _path: list[vy_ast.Module] = field(default_factory=list)
+
+    def push_path(self, module_ast: vy_ast.Module) -> None:
+        if module_ast in self._path:
+            cycle = self._path + [module_ast]
+            raise ImportCycle(" imports ".join(f'"{t.path}"' for t in cycle))
+
+        self._path.append(module_ast)
+
+    def pop_path(self, expected: vy_ast.Module) -> None:
+        assert expected == self._path.pop()
```

Comment:
Because of the side effect, I think this should be refactored into an if statement that raises an exception of some sort


The bug of missing this statement would be too insidious if the assert was somehow skipped, no matter how unlikely that would be